packages/jupyterlab-manager/src/plugin.ts

```diff
@@ -1,101 +1,236 @@
 } from '@jupyterlab/application';
 
 import {
-  IDisposable, DisposableDelegate
+  IMainMenu,
+} from '@jupyterlab/mainmenu';
+
+import {
+  RenderMimeRegistry, IRenderMimeRegistry
+} from '@jupyterlab/rendermime';
+
+import {
+  CodeCell
+} from '@jupyterlab/cells';
+
+import {
+  toArray, filter
+} from '@phosphor/algorithm';
+
+import {
+  DisposableDelegate
 } from '@phosphor/disposable';
 
+import {
+  AttachedProperty
+} from '@phosphor/properties';
+
 import {
   WidgetRenderer
 } from './renderer';
 
 import {
-  WidgetManager
+  WidgetManager, WIDGET_VIEW_MIMETYPE
 } from './manager';
 
 import {
   OutputModel, OutputView, OUTPUT_WIDGET_VERSION
 } from './output';
 
 import * as base from '@jupyter-widgets/base';
+
+// We import only the version from the specific module in controls so that the
+// controls code can be split and dynamically loaded in webpack.
 import {
   JUPYTER_CONTROLS_VERSION
 } from '@jupyter-widgets/controls/lib/version';
 
 import '@jupyter-widgets/base/css/index.css';
 import '@jupyter-widgets/controls/css/widgets-base.css';
 
-const WIDGET_MIMETYPE = 'application/vnd.jupyter.widget-view+json';
+const WIDGET_REGISTRY: base.IWidgetRegistryData[] = [];
 
-export
-type INBWidgetExtension = DocumentRegistry.IWidgetExtension<NotebookPanel, INotebookModel>;
+/**
+ * The cached settings.
+ */
+const SETTINGS: {saveState: boolean} = {saveState: false};
 
+/**
+ * Iterate through all widget renderers in a notebook.
+ */
+function* widgetRenderers(nb: Notebook) {
+  for (let cell of nb.widgets) {
+    if (cell.model.type === 'code') {
+      for (let codecell of (cell as CodeCell).outputArea.widgets) {
+        for (let output of toArray(codecell.children())) {
+          if (output instanceof WidgetRenderer) {
+            yield output;
+          }
+        }
+      }
+    }
+  }
+}
 
-export
-class NBWidgetExtension implements INBWidgetExtension {
-  /**
-   * Create a new extension object.
-   */
-  createNew(nb: NotebookPanel, context: DocumentRegistry.IContext<INotebookModel>): IDisposable {
-    let wManager = new WidgetManager(context, nb.rendermime);
-    this._registry.forEach(data => wManager.register(data));
-    nb.rendermime.addFactory({
-      safe: false,
-      mimeTypes: [WIDGET_MIMETYPE],
-      createRenderer: (options) => new WidgetRenderer(options, wManager)
-    }, 0);
-    return new DisposableDelegate(() => {
-      if (nb.rendermime) {
-        nb.rendermime.removeMimeType(WIDGET_MIMETYPE);
+/**
+ * Iterate through all matching linked output views
+ */
+function* outputViews(app: JupyterFrontEnd, path: string) {
+  let linkedViews = filter(
+    app.shell.widgets(),
+    w => w.id.startsWith('LinkedOutputView-') && (w as any).path === path
+  );
+  for (let view of toArray(linkedViews)) {
+    for (let outputs of toArray(view.children())) {
+      for (let output of toArray(outputs.children())) {
+        if (output instanceof WidgetRenderer) {
+          yield output;
+        }
```

Comment:
@jasongrout Just for posterity: Would you mind detailing what the plan is for replacing these rather hackish iterators?

Reply:
No specific plan yet. Basically, it would be nice to have a way in JLab to iterate through all renderers associated with a document. Before adding that to the JLab API, though, I'd love to have one more use case for this to double check that we have the right level of generality. I'm also thinking that we might find a more elegant way to do this in the future.